Fix nn.Dropout accuracy discrepancies between triton and torch implementations

EricKing626 opened the pull request:
We're proposing a PyTorch PR to address Torch Compile vs. Eager inconsistencies. The primary concern we’ve heard is accuracy divergence under Torch Compile (Triton). We targeted a concrete source of mismatch— nn.Dropout—and implemented an approach that:
1. Fuses Dropout with adjacent pointwise kernels (preserving performance), and
2. Preserves exact results with Eager mode (no accuracy regressions).

This PR introduces a compiler switch, adding torch._inductor.config.align_random_eager = True, that reproduces GPU-eager Philox consumption so dropout masks align with eager without disabling Triton fusion. 
In contrast, torch._inductor.config.fallback_random = True routes RNG ops back to eager and typically prevents fusing dropout with surrounding elementwise code, which can hurt performance. Our flag only affects RNG alignment; non-random kernels are unchanged. In practice we see parity with eager RNG and slightly better throughput than a vanilla torch.compile run where RNG may diverge from eager. Verified on MI325/MI350/MI355/H100/H200.


The dropout mask can be aligned with eager mode in FP32, FP16, BF16
<img width="405" height="156" alt="image" src="https://github.com/user-attachments/assets/646e7196-103f-4900-8092-a24a1533d891" />

Performance tested in MI325, each batch size is 8. align means our PR (align_random_eager), fallback means fallback_random
vit_b_16 baseline 6.936 ms | align 6.954 ms | fallback 7.013 ms | align/fallback 1.008x
swin_v2_t baseline 12.734 ms | align 12.773 ms | fallback 12.797 ms | align/fallback 1.002x
inception_v3 baseline 16.905 ms | align 16.903 ms | fallback 16.951 ms | align/fallback 1.003x

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos @chenyang78

EricKing626 committed:
dropout RNG fix: add custom_philox_rand.py and update replace_random.py

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

chien-an-chen committed:
implement fused random kernel, and the behavior is align to eager mode.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Merge branch 'dropout_RNG_fix' of https://github.com/ROCm/pytorch into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Inductor: RNG/dropout fixes and related updates, Add torch._inductor.config.align_random_eager as switch to our option modification

EricKing626 committed:
This PR decouples RNG from block-size meta-params to make compilation robust and RNG deterministic, by keeping BLOCK/XBLOCK out of the helper and doing any thread-round math on tid at the call site.

Problem: The custom Triton RNG helper rand_eager caused Inductor compilation failures due to a missing OpHandler schema, use of an undeclared XBLOCK in some kernels, and a 1D lane tensor that didn’t broadcast with 2D shapes (e.g., 8 vs 1024).
Solution: Refactor rand_eager to remove any XBLOCK dependency, derive per-element lanes as lane = subseq % 4 with the same shape as the index/subseq tensor (and pass only that tensor), and optionally register the OpHandler schema to silence warnings and stabilize codegen.

EricKing626 committed:
Use non-blocking custom_op and fuse multiple offsets into one kernel, then utilize lookup_seed to lookup offset

1. Replace Triton helper with a non-blocking custom_op that writes offsets directly to device tensors, avoiding any dependence on BLOCK/XBLOCK and eliminating syncs on H2D.
2. Fuse multiple offset reservations into a single kernel/op to cut launch overhead and AVOID op only generate once issue; expose a vectorized rand_eager_offsets and register_fake for meta shape inference.
3. Plumb offsets through inductor_lookup_seed (GPU-resident) so downstream RNG kernels index by offset, preserving determinism across eager and torch.compile.

EricKing626 committed:
Add VEC meta-parameter to align RNG vectorization; enable fp16/bf16 parity with eager

This PR introduces a VEC compile-time parameter to the Triton rand_eager path and plumbs it through codegen/lowering. VEC controls RNG consumption and grouping (4 for fp32, 8 for fp16/bf16, 2 for fp64), matching PyTorch eager’s Philox/curand_uniform4 behavior. We pass VEC as a keyword meta-arg, keep comparisons in FP32, and use the tid with per-round offset advance. As a result, dropout masks for fp16/bf16 now bit-align with eager CUDA, without changing the kernel’s random API shape for other dtypes.

EricKing626 committed:
Remove custom_philox_rand.py and change name of custom_op

EricKing626 committed:
Merge remote-tracking branch 'upstream/main' into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Update replace_random.py

EricKing626 committed:
Create test_dropout_align_random_eager.py

EricKing626 committed:
Merge branch 'dropout_RNG_fix' of https://github.com/ROCm/pytorch into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Add comment on ops_handler.py

EricKing626 committed:
Add make_prim inductor_random defination with align_dtype

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Make the triton and triton_helpers(kernel) different name

EricKing626 committed:
Merge remote-tracking branch 'origin/dropout_RNG_fix' into dropout_RNG_fix

EricKing626 committed:
Add pytest format to test_dropout_align_random_eager.py

EricKing626 commented:
@pytorchbot  drci

EricKing626 committed:
Add codegen/cpp.py

EricKing626 committed:
Merge branch 'main' into dropout_RNG_fix

EricKing626 committed:
Add halide for support, but not fully test; cancel triton import

EricKing626 committed:
Fix lint error in HL

EricKing626 commented:
@pytorchbot drci

EricKing626 committed:
Fix lint errors

EricKing626 committed:
Fix all lint errors

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Apply lintrunner

EricKing626 commented:
@pytorchbot drci

EricKing626 committed:
Restore .lintrunner.toml to upstream/main version

eellison commented on test/inductor/test_dropout_align_random_eager.py:
```diff
@@ -0,0 +1,252 @@
+# Tests: correctness (forward/backward)
+# ───────────────────────────────────────────────────────────────
+@pytest.mark.parametrize("training", [False, True])
+def test_linear_block_compile_parity_forward(device, input_tensor, training):
```
Can you please copy the test inductor idioms for test classes - see:

https://github.com/pytorch/pytorch/blob/aa4a8c9b9228a8999cac1fba37cd04e086ecfb9b/test/inductor/test_control_deps.py#L2-L16

eellison commented on test/inductor/test_dropout_align_random_eager.py:
```diff
@@ -0,0 +1,252 @@
+# ───────────────────────────────────────────────────────────────
+class LinearBlock(torch.nn.Module):
+    def __init__(self, hidden_dim: int, ffn_dim: int, dropout: float = DROPOUT_P):
+        super().__init__()
```
Can we also get tests for :

a) dynamic shapes (you can use mark_dynamic or dynamic=True)
b) cudagraphs - you can invoke with mode='reduce-overhead'

EricKing626 replied:
Dynamic_shape CI tests can be pass but I found it doesn't update new dropout when inter-step training because set_offset is host API and require concrete integer. So, I turn off it when dynamic_shape.

Dynamic shape related:
python test/inductor/test_fused_attention.py SDPAPatternRewriterGpuDynamicTests.test_sdpa_rewriter_12_gpu
python test/inductor/test_aot_inductor.py AOTInductorTestABICompatibleGpu.test_constant_int_kernel_input_cuda
pytest test/inductor/test_dropout_align_random_eager.py -k test_dropout_parity_dynamic_shapes

EricKing626 replied:
Added

eellison reviewed (commented):
Some questions from me. also cc @v0i0 re: inductor/eager numerics.

eellison commented on torch/_inductor/fx_passes/replace_random.py:
```diff
@@ -20,6 +20,70 @@
+    dev_index = device.index if isinstance(device, torch.device) else int(device)
+    gen = torch.cuda.default_generators[dev_index]
+    old_off = int(gen.get_offset())
+    gen.set_offset(old_off + used_offset)
```
And I don't think set_offset will work with cudagraph

EricKing626 replied:
I got your point. But is there any better place to reserve and push the RNG offset for supporting cudagraph in your opinion?

EricKing626 committed:
Update the advise from eellison

EricKing626 committed:
Avoid test case going into cpu

EricKing626 committed:
Update test_dropout_align_random_eager.py

EricKing626 committed:
Make RNG seed a runtime argument instead of reading it during lowering, so the seed is not baked as a constant. Align compiled dropout RNG with eager by passing a packed RNG state at runtime.

Problem: initial_seed() was read in lowering, causing the seed to be constant-folded and ignoring torch.manual_seed / user generators at runtime.

Fix:

Refactor helper to compute only threads_per_round; no RNG access in lowering.

Introduce runtime RNG state ([seed, base_offset]) returned by a custom op; pass this state as the second arg to inductor_prims.random.

Update lowering to unpack [seed, base_offset] (e.g., seed_loader([0]), seed_loader([1])) and compute element offsets there.

Update replace_random and the fuse pass to batch-create RNG states (rand_eager_states) and select rows via aten.select (no lookup_seed for scalars).

Result:

Seeds respect runtime changes (manual_seed, per-device generators).

Supports multiple dropouts per iteration and multi-iteration parity (distinct non-overlapping streams via offsets).

Avoids FX Proxy in control flow; no seed baking in compiled graphs.

Tests:

Parity for eager vs compiled (single/multi-dropout, dynamic shapes, cudagraphs).

Multi-iteration parity with BASE_SEED + i.

EricKing626 committed:
Fix Inductor RNG alignment on CPU by avoiding torch.cuda in CPU paths and registering custom RNG ops as fallback to avoid missing lowering errors.

1. Add device checks in inductor_random / get_threads_per_round so CPU paths no longer call any torch.cuda.*, fixing CPU / Halide / Triton-CPU dropout and gumbel-softmax tests (including environments without CUDA or GPU drivers).

2. Register custom_op::rand_eager_offset and custom_op::rand_eager_offsets via make_fallback, so these RNG helper ops run through their Python implementation instead of requiring a dedicated lowering, resolving MissingOperatorWithoutDecomp errors in CUDA bernoulli/dropout tests.

EricKing626 commented:
@pytorchbot drci

EricKing626 committed:
Add dummy for ensure not using torch.cuda for CPU or XPU cases

EricKing626 commented:
@pytorchbot drci

EricKing626 committed:
Fix inductor_random device handling

Use torch.device in get_threads_per_round to safely derive CUDA properties or fall back to a CPU grain size, preventing the NameError in GPUTests.test_bernoulli2_cuda without changing existing RNG semantics.

EricKing626 committed:
Add lowering CPU guard to prevent cpu path, add pytest in test_dropout_align_random_eager

EricKing626 committed:
Restore Triton dropout lowering by reverting the lookup_seed–based RNG state wiring and switching to a packed (seed, offset) int64 representation that keeps GPUTests.test_dropout2_cuda passing.

The previous revision tried to route fused RNG state through inductor_prims.lookup_seed plus aten.select.int on a [N, 2] tensor. While this worked functionally, it broke the Triton RNG pattern matcher: the dropout lowering no longer recognized the RNG seed source, so no tl.rand call was emitted and GPUTests.test_dropout2_cuda started failing (fw_code.count("tl.rand") == 0).

This PR changes the approach back to something Triton can reliably match:

Each (seed, offset) pair is encoded as a single 64-bit integer:

packed = (seed.to(torch.int64) << 32) + offset

high 32 bits = seed, low 32 bits = offset

fuse_offset_creation_pass now builds a 1D int64 tensor of these packed states for all custom_op::rand_eager_offset calls on a device, and each original node is replaced by an aten.select.int on that packed tensor.

In the Triton lowering, we decode back to the original state:

rng_seed = packed >> 32

base_offset = packed & 0xFFFF_FFFF

This keeps RNG state creation horizontally fused (reducing launch overhead), but from the lowering’s point of view the seed/offset still flow through a simple, recognizable pattern that produces a single tl.rand in the generated kernel.

Semantics of align_random_eager and RNG alignment vs eager mode are unchanged; only the internal representation of (seed, offset) is adjusted so that:

Triton RNG pattern matching works again.

python test/inductor/test_torchinductor.py GPUTests.test_dropout2_cuda -k dropout2 passes.

EricKing626 committed:
Fix CI errors and simplify custom_op::rand_eager_offsets

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Update replace_random.py

EricKing626 committed:
Guard RNG host-state ops from CUDA Graph capture; add CPU barrier & device-side packing for rand_eager_offsets

This PR fixes CUDA Graph capture failures triggered by RNG host-state queries in our random replacement path.
Problem
When rand_eager_offset(s) runs inside a captured region, it calls CUDAGeneratorImpl::initial_seed() / get_offset() / set_offset(), which are illegal during CUDA Graph capture, causing:
RuntimeError: Cannot call CUDAGeneratorImpl::current_seed during CUDA graph capture.
What this PR changes (minimal surface)

Exclude RNG host-state ops from capture:

Insert a tiny CPU barrier custom op (custom_op::cpu_barrier) that executes in eager and returns a non-aliasing clone, creating a graph boundary before rand_eager_offsets.

Mark fused offsets node with node.meta["inductor_no_cudagraphs"] = True so Inductor won’t cudagraphify the subgraph containing RNG state reads.
Result

RNG seed/offset reservation happens outside capture; the captured region contains only device-safe ops.

Eliminates the cudagraph error above and keeps the random path deterministic/aligned.

Tests unblocked

GPUTests.test_dropout_deterministic_cuda now passes with triton.cudagraphs=True.

(Related dropout tests that rely on the same RNG path also pass locally, but the primary regression this addresses is test_dropout_deterministic_cuda.)
Result

RNG seed/offset reservation happens outside capture; the captured region contains only device-safe ops.

Eliminates the cudagraph error above and keeps the random path deterministic/aligned.

Tests unblocked

GPUTests.test_dropout_deterministic_cuda now passes with triton.cudagraphs=True.

(Related dropout tests that rely on the same RNG path also pass locally, but the primary regression this addresses is test_dropout_deterministic_cuda.)

EricKing626 committed:
Change pin_memory and non_blocking copy to accelerate custom_op

EricKing626 committed:
Update replace_random.py

EricKing626 committed:
Add primitive random function tests

EricKing626 committed:
Add cudagraph_unsafe tags and fix lint errors

EricKing626 committed:
Disable align_random_eager for dynamic shapes in random replace pass

This change restricts the align_random_eager path in replace_random.py to cases where the random size is fully static (all dimensions are plain int). When size contains SymInt (e.g., tensors marked with torch._dynamo.mark_dynamic), we fall back to the original replacement function.

The reason is that _shape_to_offset(size, device) currently computes a concrete offset by iterating over size and calling int(s). For dynamic shapes, this would forcibly specialize symbolic dimensions (such as the batch dimension in SDPA tests) to their hint values (e.g., 4), which conflicts with ShapeEnv’s dynamic-shape constraints and can trigger ConstraintViolationError.

With this guard:

Static-shape kernels still benefit from align_random_eager and fused RNG behavior.

Dynamic-shape workloads keep the existing Inductor random lowering, avoiding accidental specialization of SymInt dimensions and preserving ShapeEnv invariants.

EricKing626 committed:
Make rand_eager_offsets custom op visible to Inductor subprocess benchmarks

Problem

The TestTritonWrapper.test_wrapper_using_gpu_seed test invokes a compiled Inductor benchmark module in a subprocess via subprocess.check_output.

In the parent process, the compiled module runs correctly and uses torch.ops.custom_op.rand_eager_offsets.

In the subprocess, importing the compiled module fails with
AttributeError: '_OpNamespace' 'custom_op' object has no attribute 'rand_eager_offsets'.

Root Cause

The rand_eager_offsets custom op was registered only as a side effect of importing torch._inductor.fx_passes.replace_random.

The Inductor runtime scripts used by the subprocess import torch._inductor but do not import replace_random, so the custom op registration never ran in the child process.

As a result, the subprocess cannot resolve torch.ops.custom_op.rand_eager_offsets even though the parent process can.

Fix

Ensure that importing torch._inductor always triggers registration of the rand_eager_offsets custom op by importing the FX pass module from torch._inductor.__init__:

Add from .fx_passes import replace_random as _inductor_replace_random # noqa: F401

This keeps the import as a pure side effect for registration while avoiding unused-import lint errors.

With this change, both the parent process and any subprocess that imports torch._inductor see the same custom op registry.

Testing

Re-ran test_triton_wrapper.py::TestTritonWrapper::test_wrapper_using_gpu_seed to confirm that:

The compiled module runs correctly in the parent process.

The subprocess can import and execute the compiled module without raising AttributeError.

Verified that Inductor still compiles and runs the dropout path using rand_eager_offsets as expected.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Replace Dropout RNG Custom Ops with Inductor Prims to Enable Subprocess Execution

This PR replaces the previous custom_op–based RNG alignment mechanism with Inductor prims (inductor_rand_eager_offset and inductor_rand_eager_offsets).
Custom ops were not visible to subprocesses, causing failures in test_triton_wrapper.py::TestTritonWrapper.test_wrapper_using_gpu_seed.

By migrating to Inductor prims:

RNG state reservation is fully handled inside Inductor’s primitive system.

Subprocess-executed compiled modules no longer fail due to missing custom ops.

Dropout RNG alignment behavior remains unchanged and fully compatible with Inductor fusion and lowering flows.

This resolves all subprocess execution issues and allows
python test/inductor/test_triton_wrapper.py -k test_wrapper_using_gpu_seed
to pass consistently.

EricKing626 committed:
Preserve SymInt shapes in _shape_to_offset for dynamic RNG alignment

This PR fixes a dynamic-shape bug in the RNG alignment path by avoiding premature concretization of symbolic dimensions. Specifically, _shape_to_offset now multiplies by s directly instead of int(s), so nelem remains a SymInt when shapes are dynamic. This keeps ShapeEnv constraints consistent, prevents guard ConstraintViolationError for dynamic batch sizes, and restores correct behavior for align_random_eager under dynamic shapes.

EricKing626 committed:
Fix lint errors

EricKing626 committed:
Disable dynamic_shape support because inter-graph will have issue because the global host state does not advance, causing identical random sequences across repeated runs.

Only enable align_random_eager for fully static shapes.
If `size` contains SymInt (dynamic dims), forcing it through _shape_to_offset
would specialize those symbolic dimensions to concrete values and violate
ShapeEnv/dynamic-shape constraints, so we fall back to the original replacement.
If remove the constrain of replace_random.py L204 of all(isinstance(s, int) for s in size), all CI tests will pass but inter-graph will have issue because the global host state does not advance, causing identical random sequences across repeated runs.
The Fundamental Conflict: Violation of Causality and LocalityLocality (Space):The generator state resides in Host Memory (CPU), but the dynamic shape value ($u0$) is computed and stored in Device Memory (GPU).Causality (Time):PyTorch relies on Asynchronous Dispatch. The CPU launches kernels before the GPU executes them.gen.set_offset is a Host instruction requiring a concrete value now (Launch Time).$u0$ is a Device result produced later (Execution Time).Principle: The Host cannot consume a future value that has not yet been computed.The Cost of Resolution (Synchronization):Bridging this gap requires a Host-Device Sync (copying $u0$ back to CPU). This forces the CPU to wait, destroying the Latency Hiding mechanism that makes GPU computing fast.Conclusion:We disable dynamic shape support for set_offset because satisfying its data dependency would introduce a synchronization barrier, fundamentally negating the performance benefits of asynchronous execution.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Fix lint error

EricKing626 committed:
Merge branch 'dropout_RNG_fix' of https://github.com/ROCm/pytorch into dropout_RNG_fix

EricKing626 committed:
Fix lint error

EricKing626 committed:
Update test_dropout_align_random_eager.py

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
[Inductor/CUDA] Add inductor_reserve_rng_state op for graph-safe RNG offset management

Summary: This PR introduces a new native operator aten::inductor_reserve_rng_state to support make_graphed_callables with Inductor's random number generation.
Motivation: Currently, Inductor manages RNG offsets in Python using gen.get_offset() and gen.set_offset(). This approach fails during CUDA Graph capture because:
	1. set_offset is an eager CPU operation and is captured as a constant, preventing offset advancement during graph replay.
	2. Calling philox_engine_inputs directly is forbidden during capture (assertNotCapturing).
Changes:
	1. Implemented inductor_reserve_rng_state in CUDAGeneratorImpl.cpp.
		○ Graph Mode: Manually calls state_->increase() to advance the intragraph offset and returns the seed_extragraph_ and offset_extragraph_ pointers (tensors). These tensors are automatically updated by replay_prologue during execution.
		○ Eager Mode: Returns concrete seed/offset values wrapped in 1D tensors to maintain shape consistency (1,).
	2. Exposed private state_ access to this function via friend declaration.
Impact: This allows Inductor to generate graph-safe RNG code that correctly advances random states across multiple graph replays, fixing correctness issues in make_graphed_callables.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Update native_functions.yaml

EricKing626 committed:
[Inductor] Fix build errors and disable auto-binding for inductor_reserve_rng_state

Adds a forward declaration to resolve friend class visibility and sets manual_cpp_binding: True to bypass incorrect Python binding generation.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Fix type error for repletely call torch.ops.inductor_prims.inductor_reserve_rng_state()

EricKing626 committed:
[Inductor] Move inductor_reserve_rng_state from aten to inductor_prims and make it graph-safe

What:
Introduce torch.ops.inductor_prims.inductor_reserve_rng_state(generator?, increment)
to reserve Philox RNG state using CUDAGeneratorImpl::philox_cuda_state().

Replace the previous call path from
torch.ops.aten.inductor_reserve_rng_state to
torch.ops.inductor_prims.inductor_reserve_rng_state.

Why

Inductor needs a graph-safe way to reserve RNG offsets during CUDA/HIP graph
capture without relying on Python-managed seed/offset.

philox_cuda_state() provides the correct eager-like semantics:
in capture it returns extragraph pointers plus intragraph offset, and in eager
it returns concrete (seed, offset).

How:
Register the operator in inductor_prims with CUDA/HIP + BackendSelect
dispatch so it works even when the operator has no tensor arguments.

Behavior

Eager: returns 1-element tensors for seed/offset; intragraph offset is 0.

Graph capture: returns device pointers to extragraph seed/offset tensors and a
CPU intragraph offset value, enabling replay to advance RNG without freezing.

EricKing626 committed:
Inductor: expose inductor_reserve_rng_state as an inductor_prims op for CUDA/HIP graphs

This PR introduces torch.ops.inductor_prims.inductor_reserve_rng_state, a small C++ Inductor op that reserves Philox RNG state via CUDAGeneratorImpl::philox_cuda_state() and returns (seed, offset, intragraph_offset).

Enables CUDA/HIP graph capture + replay to use graph-safe generator state (ptr-based extragraph tensors during capture, value-based tensors in eager).

Registers BackendSelect, CUDA, and HIP kernels so the op works even when inputs contain no Tensors.

Updates Inductor to call the new inductor_prims op instead of the previous torch.ops.aten.inductor_reserve_rng_state.

Added BackendSelect.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Add the TORCH_LIBRARY_FRAGMENT back

Accidentally removed four functions from the existing TORCH_LIBRARY_FRAGMENT definitions. Add them back and fix some lint warnings.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Fix inductor_reserve_rng_state registration by requiring a Generator argument

This PR fixes build failures in torch/csrc/inductor/inductor_ops.cpp by aligning the inductor_reserve_rng_state kernel signature with PyTorch’s supported registration types.
Since the Python caller always passes torch.cuda.default_generators[device], we make the Generator argument required (remove the optional form) and update the schema accordingly. This avoids optional-type inconsistencies across builds and ensures the op registers cleanly on CUDA/HIP/BackendSelect.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Enable Dynamic-Shape RNG Offset Reservation for Inductor Philox State

This PR updates Inductor’s RNG state reservation path to support dynamic shapes by accepting a symbolic used_offset (SymInt) end-to-end instead of forcing an eager int() cast. This avoids prematurely concretizing offsets during tracing/compilation, allowing AOTAutograd / torch.compile RNG functionalization to compute shape-dependent Philox consumption correctly at runtime while preserving CUDA/HIP graph safety and existing eager behavior.

EricKing626 committed:
Inductor: register inductor_reserve_rng_state with non-optional Generator and guard CPU builds

This PR adds a CUDA/ROCm-only implementation of inductor_prims::inductor_reserve_rng_state that takes a non-optional Generator and a SymInt increment, returning (seed, offset, intragraph_offset) tensors for Inductor RNG state reservation.

Key points:

Registers the op schema in inductor_prims with Generator (not Generator?) to match call sites that always pass a concrete generator.

Provides a CUDA/ROCm implementation using CUDAGeneratorImpl::philox_cuda_state() and returns pointer-backed tensors during graph capture.

Adds a safe CPU stub (guarded build) that fails fast with a clear error message if called, preventing accidental CPU linkage to CUDA/ROCm symbols and avoiding undefined references in CPU-only link steps.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Guard inductor_reserve_rng_state registration to avoid CPU-only linker issues

This PR gates the inductor_prims::inductor_reserve_rng_state schema/registration behind USE_CUDA || USE_ROCM so CPU-only builds never reference CUDA generator symbols or expose the op for accidental use. CUDA/ROCm builds continue to register the prim and keep torch.ops.inductor_prims.inductor_reserve_rng_state(...) working as before.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
[Inductor] Fix dynamic shapes UT issues and linker error by decoupling CUDA RNG implementation from CPU ops

This PR resolves an undefined reference linker error in libtorch_cpu.so by separating the inductor_reserve_rng_state implementation from its schema registration.
Move the inductor_reserve_rng_state op schema and registration out of inductor_ops.cpp into a CUDA/ROCm-only source file so CPU-only targets never see the schema or pull in CUDA generator symbols, while GPU builds continue to expose torch.ops.inductor_prims.inductor_reserve_rng_state.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Change lambda to function def; include inductor_ops.h

EricKing626 committed:
Merge branch 'dropout_RNG_fix' of https://github.com/ROCm/pytorch into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Fix Inductor RNG reserve op build guards and ATen header includes

This PR moves/guards inductor_reserve_rng_state so CPU-only targets don’t see the op schema or pull in CUDA generator symbols, avoiding linker failures. For CUDA/ROCm builds, the op remains registered and works as before. It also adds the minimal per-op ATen headers needed under AT_PER_OPERATOR_HEADERS to fix missing at::scalar_tensor/zeros declarations.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Different function name to seperate inductor_reserve_rng_state schema and function

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Split Inductor RNG reserve op into CPU schema + GPU implementation

Move inductor_prims::inductor_reserve_rng_state operator schema to the CPU/common Inductor ops file and keep the CUDA/HIP kernel implementation in inductor_ops_gpu.cpp. This avoids link/registration issues in non-CUDA targets and keeps GPU-specific includes isolated.

EricKing626 committed:
Fix zero division error and CUDA Graph capture pinned-CPU copy error

| Repro command | Failure (where) | Root cause | Fix / mitigation |
|---|---|---|
| PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=2 python test/inductor/test_torchinductor_opinfo.py TestInductorOpInfoCUDA.test_comprehensive_bernoulli_cuda_float64 | replace_random.py::_shape_to_offset() → ZeroDivisionError | The sample input contains an empty tensor (0, 3), so nelem=0 → grid_size=0 → division by zero in (nelem - 1) // (block_size * grid_size * unroll). | Add an early guard in _shape_to_offset: if nelem == 0: return 0 (or otherwise ensure the denominator can’t become 0). |
| PYTORCH_TEST_WITH_ROCM=1 python test/inductor/test_cudagraph_trees.py CudaGraphTreeTests.test_cached_forward_backward | CUDA Graph capture → RuntimeError: Cannot copy between CPU and CUDA tensors during CUDA graph capture unless the CPU tensor is pinned | During graph capture, a CPU↔GPU copy happens (e.g., intra_t.to(device=...) inside _reserve_rng_state), and the CPU tensor is not pinned. CUDA graph capture forbids such copies unless CPU memory is pinned. | Avoid CPU tensors / .to("cuda") inside capture: move all device transfers outside capture, keep intra_t on GPU (or return pinned-CPU intra_t), and ensure _reserve_rng_state does not trigger CPU↔GPU copies during capture. |

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Fix error in test_aot_inductor.py::AOTInductorTestABICompatibleGpu::test_constant_int_kernel_input_cuda

This failure is caused by a data-dependent SymNode guard introduced by if nelem == 0 in _shape_to_offset
Fix by replacing the Python branch with dynamic-shape friendly guards (guard_or_false / statically_known_true) to avoid GuardOnDataDependentSymNode while still handling empty tensors.

EricKing626 committed:
Merge branch 'dropout_RNG_fix' of https://github.com/ROCm/pytorch into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Remove test_dropout2 mps xfail

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Remove MPS xfail in test_dropout3

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Add back MPS xfail in test_dropout3

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Fix test_dropout2 & dynamic-shapes test_dropout3 on MPS by forcing the existing xfail path when align_random_eager is enabled

Fix dynamic-shapes dropout3 on MPS by forcing the existing xfail path when align_random_eager is enabled

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 commented:
> @EricKing626 great! we'll take a look.

@eellison Can we re-run failed tests? This all just flaky errors. Or is leaving these flaky errors ok?
And another thing, do we need to close the flag as default when merging?
<img width="599" height="308" alt="image" src="https://github.com/user-attachments/assets/72b4083d-d172-4519-b0c0-771c2e175c8f" />

EricKing626 reviewed (commented):
Functionality is already complete. Performance improves compared to the fallback_random implementation.

eellison commented on test/inductor/test_dropout_align_random_eager.py:
```diff
@@ -0,0 +1,439 @@
+    # ───────────────────────────────────────────────────────────
+    # Helper for primitive random parity (rand / randn / randint)
+    # ───────────────────────────────────────────────────────────
+    def _run_primitive_random_parity(self, kind, device, shape):
```
Here were some additional failing tests (helped find with claude). https://gist.github.com/eellison/88aa2bb71f64cc01d4f5f4b2a4fc4e8a

- float64 dtype vec mismatch
- Large seed (>32-bit) packing truncation
- Non-contiguous sliced
- numel % 4 != 0 produces wrong masks

EricKing626 replied:
First 2 fixed with added test cases in test_dropout_align_random_eager.py
"Non-contiguous sliced" & "numel % 4 != 0 produces wrong masks" might be fallback to eager

eellison reviewed (commented):
Looks great, thanks for your work here ! I left a comment on some remaining cases that don't yet work. 

Having said that, this is already useful as is, and the pr is getting large. What do you think about explicitly covering the cases we do not yet handle, and falling back to eager ? We could also config this option so we lower and warn, or always lower. 

What do you think?

EricKing626 committed:
Fix float64 dtype vec mismatch & large seed (>32-bit) packing truncation

Fix float64 dtype vec mismatch
Fix large seed (>32-bit) packing truncation
Add at::Tag::pt2_compliant_tag
Remove default align_dtype from torch/_inductor/fx_passes/replace_random.py
Test cases:
 1. make .5, just to maximally test divergence
 2. using atol=0, rtol=0 across tests

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 commented on torch/_inductor/fx_passes/replace_random.py:
```diff
@@ -29,10 +64,59 @@ def replace_random_passes(gm: torch.fx.GraphModule):
+        b = inductor_lookup_seed(offsets, 1)
+    We do this because seed creation is entirely launch overhead bound.
+    """
+    device_offsets = collections.defaultdict(list)
```
Thanks for the nit! This pass is modeled after fuse_seed_creation_pass, which also uses collections.defaultdict(list) without a local annotation. I want to keep the local defaultdict un-annotated to avoid overly specific/brittle typing.

EricKing626 committed:
Add tests of float64 dtype vec & mismatch Large seed (>32-bit) packing truncation

EricKing626 committed:
Fix Lint Error

jansel reviewed (commented):
This looks great!  Two things:
1) Do we have any perf data on this?  If it is faster maybe this should be the default rather than behind a flag.
2) Looks like there are test failures.

EricKing626 committed:
Update test_dropout_align_random_eager.py

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Update test_dropout_align_random_eager.py

This PR relaxes the strict half-precision parity check, improving flexibility for future PyTorch half-precision behavior changes and reducing the risk of CI failures from minor mask differences.

EricKing626 committed:
Tighten dropout parity criteria for half-precision tests

This PR makes the test criteria stricter by enforcing a tighter tolerance for eager vs. compiled half-precision dropout parity.

EricKing626 committed:
Merge branch 'pytorch:main' into dropout_RNG_fix

EricKing626 committed:
Update test_dropout_align_random_eager.py

EricKing626 committed:
Merge branch 'dropout_RNG_fix' of https://github.com/ROCm/pytorch into dropout_RNG_fix

EricKing626 committed:
Fix lint error

EricKing626 committed:
Move test_float64 into test_dropout_mask_parity_and_rng_offset_cuda

EricKing626 committed:
Update test_dropout_align_random_eager.py

Remove float64 support

eellison commented on torch/csrc/inductor/inductor_ops_gpu.cpp:
```diff
@@ -0,0 +1,89 @@
+// by explicitly handling the internal generator state via public APIs.
+//
+// Behavior:
+// - Graph Mode: Advances the generator state and returns pointers (wrapped as
```
cc maybe @eee4017 for comments

eellison reviewed (approved):
Awesome! cc @albanD , do we have any tests we do to verify random distributions, I know you recently landed some changes here. 

Before we land, lets do a dashboard run to double check that there arent any failures or regressions.

Also, I see that this is on by default, and does not warn if it's one of the cases we have not yet handled (Non-contiguous sliced, numel % 4 != 0 ).

How are you envisioning the use case ? would it make sense to add an option to either warn or fallback to eager if we can't prove we'll match numerics ?

albanD reviewed (commented):
@eellison if the new algorithm is bitwise equal with eager, I guess you can simply compare with eager given a set seed?

If you want to check the statistical properties of a new op, we use scipy.stats. See test_torch.py for how we use it.

EricKing626 committed:
Turn align_random_eager off default
